src/node_report_module.cc

```diff
@@ -0,0 +1,506 @@
+static uv_mutex_t node_isolate_mutex;  // mutex for watchdog thread
+static struct sigaction saved_sa;  // saved signal action
+#endif
+
+// State variables for v8 hooks and signal initialisation
+static bool exception_hook_initialised = false;
+static bool error_hook_initialised = false;
+static bool signal_thread_initialised = false;
+
+static Isolate* node_isolate;
+
+/*******************************************************************************
+ * External JavaScript API for triggering a report
+ *
+ ******************************************************************************/
+void TriggerReport(const FunctionCallbackInfo<Value>& info) {
+  Isolate* isolate = info.GetIsolate();
+  HandleScope scope(isolate);
+  char filename[NR_MAXNAME + 1] = "";
+  MaybeLocal<Value> error;
+  int err_index = 0;
+
+  if (info[0]->IsString()) {
+    // Filename parameter supplied
+    String::Utf8Value filename_parameter(isolate, info[0]);
+    if (filename_parameter.length() < NR_MAXNAME) {
+      snprintf(filename, sizeof(filename), "%s", *filename_parameter);
+    } else {
+      isolate->ThrowException(String::NewFromUtf8(
+                              isolate,
+                              "node-report: filename parameter is too long",
+                              NewStringType::kNormal).ToLocalChecked());
+    }
+    err_index++;
+  }
+
+  // We need to pass the JavaScript object so we can query it for a stack trace.
+  if (info[err_index]->IsNativeError()) {
+    error = info[err_index];
+  }
+
+  if (nodereport_events & NR_APICALL) {
+    TriggerNodeReport(isolate, kJavaScript,
+                      "JavaScript API",
+                      __func__,
+                      filename,
+                      error);
+    // Return value is the report filename
+    info.GetReturnValue().Set(String::NewFromUtf8(isolate, filename));
+  }
+}
+
+/*******************************************************************************
+ * External JavaScript API for returning a report
+ *
+ ******************************************************************************/
+void GetReport(const FunctionCallbackInfo<Value>& info) {
+  Isolate* isolate = info.GetIsolate();
+  HandleScope scope(isolate);
+  std::ostringstream out;
+
+  MaybeLocal<Value> error;
+  if (info[0]->IsNativeError()) {
+    error = info[0];
+  }
+
+  GetNodeReport(isolate, kJavaScript, "JavaScript API", __func__, error, out);
+  // Return value is the contents of a report as a string.
+  info.GetReturnValue().Set(String::NewFromUtf8(isolate, out.str().c_str()));
+}
+
+/*******************************************************************************
+ * External JavaScript APIs for node-report configuration
+ *
+ ******************************************************************************/
+void SetEvents(const FunctionCallbackInfo<Value>& info) {
+  Isolate* isolate = info.GetIsolate();
+  String::Utf8Value parameter(isolate, info[0]);
```

Comment:
For all of these functions, either:

- Make sure that this doesn’t fail before calling the C++ wrapper, and add `CHECK_NOT_NULL(*parameter);` afterwards; or:
- Return to JS if `*parameter` is `nullptr` instead of crashing